### ISSUE TITLE
Pylint fails to detect NameError from type hint

### ISSUE THREAD

**Issue by turettn** [PyCQA/pylint#3771]
### Steps to reproduce

Pylint the following code:
```python3
import math


class Point:
    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y

    def distance_to(self, other: Point) -> float:
        x_diff = self.x - other.x
        y_diff = self.y - other.y
        return math.sqrt(x_diff ** 2 + y_diff ** 2)


def main() -> None:
    origin = Point(0, 0)
    pnt = Point(10, 15)
    print(f"Distance: {pnt.distance_to(origin)}")


if __name__ == "__main__":
    main()
```

### Current behavior

Pylint is perfectly happy:

```
(venv) [nturett@dev core]$ pylint test.py
(venv) [nturett@dev core]$ 
```

Python is not:

```
(venv) [nturett@dev core]$ python test.py
Traceback (most recent call last):
  File "test.py", line 4, in <module>
    class Point:
  File "test.py", line 9, in Point
    def distance_to(self, other: Point) -> float:
NameError: name 'Point' is not defined
```

### Expected behavior

Since this code will apparently never run, I would expect that pylint would return an undefined-variable error, matching the interpreter's actions.

Interestingly, if I change the line to include a misspelling, pylint catches it:

```python
    def distance_to(self, other: Piont) -> float:
```

yields:

```
test.py:9:33: E0602: Undefined variable 'Piont' (undefined-variable)
```

### pylint --version output

```
(venv) [nturett@dev core]$ pylint --version
pylint 2.5.3
astroid 2.4.2
Python 3.7.8 (default, Jul 24 2020, 20:26:49) 
[GCC 7.3.1 20180712 (Red Hat 7.3.1-9)]
```
